contracts/WETH10.sol

```diff
@@ -6,10 +6,14 @@ pragma solidity 0.7.0;
 import "./IWETH10.sol";
 
 
-interface ERC677Receiver {
+interface TransferReceiver {
     function onTokenTransfer(address, uint, bytes calldata) external;
 }
 
+interface ApprovalReceiver {
```

Comment:
I am personally not a fan of adding specific method to receive approvals.
this add complexity on the receiver side and require them to add special public function for it

I think a better approach is, like https://eips.ethereum.org/EIPS/eip-2771 to append the needed data at the end of the call data.

This allow receiver to keep using their method that they already use for when the sender is using the approve and transfer pattern. They just need to check the appended data which can be easily extracted from the appended data.

For an approval the only needed information is actually the sender address (the amount is not required and do not add security) and so it can be exactly like eip-2771: `_getMsgSender`

Reply:
If I read [EIP2771](https://eips.ethereum.org/EIPS/eip-2771) right, the receiver has to:
 - Implement a function to replace calls to `msg.sender`
 - Using assembly
 - After implementing a whitelist of trusted forwarders

That is complex, but I like part of the idea.

I wouldn't mind replacing the `onTokenApproval` call for a `spender.call(data)` call instead, without appending `msg.sender` to the data. The receiver should know how do deal with it even if `msg.sender` is the WETH contract.

If we do that, though, we would have to replace `onTokenTransfer` as well for a `to.call(data)`, for consistency. That would make us depart even further from EIP667, but that's not something that really bothers me.

Reply:
the trusted forwarder part is part of eip2771 not of what I propose here

the assembly part is probably not required with modern solidity. Not sure

But for the receiver this much nicer, as you can deal with the token logic the same way in both case (the only difference is getting the sender)

Reply:
The trusted forwarder part is required whenever you have one of these callback functions. It's the same for `approveAndCall`, `transferAndCall`, and even `executeOnFlashMint`, and it is true whether you append `msg.sender` to `data` or pass it as a parameter.

All these callbacks are intended to operate on some account, with `msg.sender` being the forwarder (WETH10, in this case). The receiver needs to trust the forwarder on the account that originated the transaction, otherwise anyone could call the callback targeting any account.

So in that sense, the receiver already has to do some work in establishing that trust, either with the `_msgSender` function or with a whitelist on `onTokenApproval`.

Considering that, your proposal looks more elegant, but now my concern is that if we remove the `onTokenApproval` interface, for consistency, we would have to remove the `onTokenTransfer` as well (while maintaining `transferAndCall` and `depositAndCall`).

Are we rushing to innovate too much, by adding this pattern that is not battle-tested in a contract that is supposed to be rock-solid?

Reply:
Yes obviously, it need to have a check for msg.sender, but in the context of my proposal saying that it require a trusted forwarder was misleading. The trusted forwarder is WETH10 itself as other solution like you said.

> Are we rushing to innovate too much, by adding this pattern that is not battle-tested in a contract that is supposed to be rock-solid?

Maybe, It depends on perspective, but it is good to at least consider it. Also when is WETH10 planned to be deployed? Do we already plan for WETH11?
if we want WETH10 to be the end goal, maybe we should spend more time ensuring we use the best solution.

Reply:
> if we remove the onTokenApproval interface, for consistency, we would have to remove the onTokenTransfer as well (while maintaining transferAndCall and depositAndCall).

I don't think this is a necessity. But we could actually have the same for transferAndCall but this time you ll have to pass more info, this is the approach taken by the more general approach: transfer-gateway : https://github.com/wighawag/transfer-gateway

Reply:
I don't think that there will be a WETH11 anytime soon. It is already an issue to fragment the liquidity between WETH9 and WETH10, and current platforms will struggle to migrate anyway. So let's get it right.

I was targeting end of year for a go live. There is only one more PR coming after the two that are currently open.

I've had a look at the transfer gateway, and I understand it much better now. I'm also working on the flash minting feature that also involves callbacks to potentially arbitrary functions so I'm seeing a pattern emerging here.

I think I can clearly see now that the issue is one of explicit callback interfaces versus appending data to `msg.data` and then depending on a specific receiver function to retrieve it.

In eip2771 the only parameter appended is `msg.sender`, in transfer gateways you append token, amount and sender. If we would use this pattern in the flash minting callback we would need sender, amount and fee (with the emerging MIP-25 standard). It is obvious that each callback will need different data extracting functions in the receiver.

I think that appending parameters to `data` and then extracting them in the receiver actually adds complexity on the side of the receiver. Adding a public function as an explicit entry point for a callback is a much more obvious pattern, with less abstraction, and less chance for unforeseen complications on the receivers.

I definitely think now that the pattern to append parameters to `msg.data` is not widespread or mature enough to be used here. However, I appreciate very much you raising it. I've learned something very interesting, and might be using the pattern in other projects.

Reply:
Fair enough,  I respect the choice you make. It is true that appending data is not mature yet, without any standard even in draft form to make it somehow official (except 2661 for metatx)

Regarding complexity on the receiver I think it actually reduce complexity and that is why I prefers over specific callback. This is because every function that need to receive payment will operate the same logic once they get hold of the sender. This is especially true for the approve flow. the functions that receive payment just need to do `transferFrom(_msgSender(), to, amount)` as usual and in all cases. No branching.